src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21083,11 +21083,199 @@ public void M()
+        field1.ToString();
+        field2.ToString();
+        field3.ToString();
+        field4.ToString(); // 1
+
+        [MemberNotNull(nameof(field1), nameof(field2))]
+        bool init<T>(T t) => throw null!;
+    }
+}
+""", MemberNotNullAttributeDefinition]);
+
+            // The ContainingSymbol on a substituted local function is incorrect (returns the containing type instead of the containing method)
+            // so we can't find the proper `this` receiver to apply the member post-conditions to.
+            // Tracked by https://github.com/dotnet/roslyn/issues/75543
+            c.VerifyDiagnostics(
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         field2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field2").WithLocation(13, 9),
+                // (15,9): warning CS8602: Dereference of a possibly null reference.
+                //         field4.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(15, 9));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_NonStaticLocalFunction_WithinInstanceMethod_Nested()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1 = "";
+    public string? field2;
+    public string field3 = "";
+    public string? field4;
+
+    public void M()
+    {
+        local();
+
+        void local()
+        {
+            init();
+            field1.ToString();
+            field2.ToString();
+            field3.ToString();
+            field4.ToString(); // 1
+
+            [MemberNotNull(nameof(field1), nameof(field2))]
+            bool init() => throw null!;
+        }
+    }
+}
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyDiagnostics(
+                // (19,13): warning CS8602: Dereference of a possibly null reference.
+                //             field4.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(19, 13));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_NonStaticLocalFunction_WithinInstanceMethod_Nested_WithinStaticLocalFunction()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1 = "";
+    public string? field2;
+    public string field3 = "";
+    public string? field4;
+
+    public void M()
+    {
+        local();
+
+        static void local()
+        {
+            init();
+            field1.ToString();
+            field2.ToString();
+            field3.ToString();
+            field4.ToString();
+
+            [MemberNotNull(nameof(field1), nameof(field2))]
+            bool init() => throw null!;
```

Comment:
It would be interesting to have versions of a few of these tests (e.g. referencing instance members in 'static context'), where the end of the local function is reachable.